R/stream_statuses.R

```diff
@@ -0,0 +1,173 @@
+    file_name = NULL,
+    append = TRUE,
+    instance = NULL,
+    token = NULL,
+    anonymous = FALSE,
+    verbose = TRUE){
+
+  path <- "/api/v1/streaming/hashtag"
+  if(isTRUE(local)){
+    path <- paste0(path,"/local")
+  }
+  params <- list(tag = hashtag)
+
+  quiet_interrupt(
+    stream_toots(timeout,file_name, append, token, path, params, instance, anonymous)
+  )
+  invisible(NULL)
+}
+
+#' @rdname stream_timeline
+#' @export
+stream_timeline_list <- function(
+    list_id,
+    timeout = 30,
+    file_name = NULL,
+    append = TRUE,
+    instance = NULL,
+    token = NULL,
+    anonymous = FALSE,
+    verbose = TRUE){
+
+  path <- "api/v1/streaming/list"
+  params <- list(list = list_id)
+
+  quiet_interrupt(
+    stream_toots(timeout,file_name, append, token, path, params, instance, anonymous)
+  )
+  invisible(NULL)
+}
+
+#' Parser of Mastodon stream
+#'
+#' Converts Mastodon stream data (JSON file) into parsed tibble.
+#' @param path Character, name of JSON file with data collected by any [stream_timeline] function.
+#' @export
+#' @seealso `stream_timeline_public()`, `stream_timeline_hashtag()`,`stream_timeline_list()`
+#' @examples
+#' \dontrun{
+#' stream_timeline_public(1,file_name = "stream.json")
+#' parse_stream("stream.json")
+#' }
+parse_stream <- function(path){
+  json <- readLines(path)
+  tbl <- dplyr::bind_rows(lapply(json,function(x) parse_status(jsonlite::fromJSON(x))))
+  tbl[order(tbl[["created_at"]]),]
+}
+
+
+stream_toots <- function(timeout,file_name = NULL, append, token, path, params,
+                                instance = NULL, anonymous = FALSE, verbose = TRUE,...){
+  if (is.null(instance) && anonymous) {
+    stop("provide either an instance or a token")
+  }
+  h <- curl::new_handle(verbose = FALSE)
+  if (is.null(instance)) {
+    token <- check_token_rtoot(token)
+    url <- prepare_url(token$instance)
+    curl::handle_setheaders(h, "Authorization" = paste0("Bearer ",token$bearer))
+  } else {
+    url <- prepare_url(instance)
+  }
+
+  if(is.null(file_name)){
+    file_name <- tempfile(pattern = "stream_toots", fileext = ".json")
+  }
+
+  url <- httr::modify_url(url,path = path,query = params)
+
+  stopifnot(is.numeric(timeout), timeout > 0)
+  stop_time <- Sys.time() + timeout
+
+  output <- file(file_name)
+  con <- curl::curl(url,handle = h)
+  open(output,open = if (append) "ab" else "b")
+  open(con = con, "rb", blocking = FALSE)
+  sayif(verbose,"Streaming toots until ",stop_time)
+  n_seen <- 0
+  while(isIncomplete(con) && Sys.time() < stop_time){
+    buf <- readLines(con,warn = FALSE)
+    if(length(buf)){
+      line <- buf[grepl("created_at",buf)] # This seems unstable but rtweet does something similar
+      line <- gsub("^data:\\s+","",line)
+      line <- complete_line(line)
+      writeLines(line,output)
+      n_seen <- n_seen + length(line)
+      cat("streamed toots: ",n_seen,"\r")
+    }
+  }
+  on.exit({
+    close(con)
+    close(output)
+  })
+  invisible()
```

Comment:
@schochastics Would it be better to return an invisible `file_name` here (the same for all `stream_*` functions)? It would be quite useful when the input `path` (or `file_name` for `stream_*`) is `NULL` (the default). If `file_name` doesn't get returned, one can probably never get the path of that temp file (unless going through all temp files).

Reply:
In cases like this:

```r
x <- stream_timeline_public(10)
```

One can never get the file back. (We might implement a `parse` parameter in the future, akin `rtweet::stream_tweets`)

Reply:
Ah i didnt quite understand what the parsing does in rtweet? Let me try to fix this. We shouldnt loose toots in tempfiles😬

Reply:
How does rtweet not loose track of the tmp file? https://github.com/ropensci/rtweet/blob/master/R/stream.R#L83-L88

Reply:
It gets echoed. If you prefer, it can also be a solution.

https://github.com/ropensci/rtweet/blob/14c89649e152f1c1609f9dc3d65234dbf2560d8e/R/stream.R#L71

Reply:
Ok i think i prefer to echo the tmp file name and not return anything. The parsing should be done when reading the json. I will push something later as a suggestion

Reply:
@schochastics Sure thing.

Reply:
@chainsawriot Ok done. tmp file is now only echoed. I encourgae users in the vignette to always set a file_name